src/main/java/io/stargate/sgv2/jsonapi/service/resolver/sort/TableSortClauseResolver.java

```diff
@@ -39,22 +44,128 @@ public TableSortClauseResolver(
   }
 
   @Override
-  public OrderByCqlClause resolve(CommandContext<TableSchemaObject> commandContext, CmdT command) {
+  public WithWarnings<OrderByCqlClause> resolve(
+      CommandContext<TableSchemaObject> commandContext, CmdT command) {
     Objects.requireNonNull(commandContext, "commandContext is required");
     Objects.requireNonNull(command, "command is required");
 
     // NOTE: Currently only supporting vector sort, next PR will deal with clustering key sorting
     var apiTableDef = commandContext.schemaObject().apiTableDef();
     var sortClause = command.sortClause();
     if (sortClause == null || sortClause.isEmpty()) {
-      return OrderByCqlClause.NO_OP;
+      LOGGER.debug("Sort clause is null or empty, no CQL ORDER BY needed.");
+      return WithWarnings.of(OrderByCqlClause.NO_OP);
     }
 
+    // All sorting can only be on columns in the table definition
+    var sortColumns = sortClause.sortColumnIdentifiers();
+    checkUnknownSortColumns(commandContext.schemaObject(), sortColumns);
+
     var vectorSorts = sortClause.tableVectorSorts();
-    if (vectorSorts.isEmpty()) {
-      // For now, only supporting vector sort
-      return OrderByCqlClause.NO_OP;
+    return vectorSorts.isEmpty()
+        ? resolveNonVectorSort(commandContext, sortClause, sortColumns)
+        : resolveVectorSort(commandContext, sortClause, vectorSorts);
+  }
+
+  /**
+   * We have atleast one sort expression, and non of them are vector sorts.
+   *
+   * <p>If the sort uses the clustering keys in the correct way according to CQL, then we can use
+   * the CQL Order By to push the sorting to the database. See {@link
+   * TableOrderByClusteringCqlClause}. If not then we return a {@link OrderByCqlClause#NO_OP} to
+   * indicate that we cannot push the sorting to the database and need to do it in memory.
+   */
+  private WithWarnings<OrderByCqlClause> resolveNonVectorSort(
+      CommandContext<TableSchemaObject> commandContext,
+      SortClause sortClause,
+      List<CqlIdentifier> sortColumns) {
+
+    var apiTableDef = commandContext.schemaObject().apiTableDef();
+
+    // If there is any sorting on non partition sorting columns, we cannot use CQL ORDER BY
+    var nonClusteringKeySorts =
+        sortColumns.stream()
+            .filter(sortColumn -> !apiTableDef.clusteringKeys().containsKey(sortColumn))
+            .sorted(CQL_IDENTIFIER_COMPARATOR)
+            .toList();
+    if (!nonClusteringKeySorts.isEmpty()) {
+      var warn =
+          WarningException.Code.IN_MEMORY_SORTING_DUE_TO_NON_PARTITION_SORTING.get(
+              errVars(
+                  commandContext.schemaObject(),
+                  map -> {
+                    map.put(
+                        "partitionSorting",
+                        errFmtApiColumnDef(apiTableDef.clusteringKeys().values()));
+                    map.put("sortColumns", errFmtCqlIdentifier(nonClusteringKeySorts));
```

Comment:
OK, see this line,  Let's say, user wants to sort on A,B.  A is not a clustering column.
in the errors.yaml, it will say
"The command sorted on the columns: ${sortColumns}". And ${sortColumns} is B.


But since in-memory sort is triggered in this case, it is actually sorting on both A and B.